From 6d807447fca92ed85d4f5d087ff148094ff76a51 Mon Sep 17 00:00:00 2001
From: ryanking13 <def6488@gmail.com>
Date: Thu, 13 Aug 2026 15:40:00 +0900
Subject: [PATCH] Use send and recv for Emscripten sockets

These changes ensure that OpenSSL's socket BIO uses the proper socket
syscalls (send/recv) instead of generic read/write operations when running
on Emscripten, Pyodide's NodeSockFS provides asynchronous socket operations
via send/recv syscalls, but read/write operations are not socket-aware.

TODO: should be upstream this to Pyodide?
---
 include/internal/sockets.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/internal/sockets.h b/include/internal/sockets.h
index c090a13629..acac9989e6 100644
--- a/include/internal/sockets.h
+++ b/include/internal/sockets.h
@@ -176,6 +176,11 @@ struct servent *PASCAL getservbyname(const char *, const char *);
 #define get_last_socket_error_is_eintr() (get_last_socket_error() == WSAEINTR)
 #define readsocket(s, b, n) recv((s), (b), (n), 0)
 #define writesocket(s, b, n) send((s), (b), (n), 0)
+#elif defined(__EMSCRIPTEN__)
+#define ioctlsocket(a, b, c) ioctl(a, b, c)
+#define closesocket(s) close(s)
+#define readsocket(s, b, n) recv((s), (b), (n), 0)
+#define writesocket(s, b, n) send((s), (b), (n), 0)
 #elif defined(__DJGPP__)
 #define closesocket(s) close_s(s)
 #define readsocket(s, b, n) read_s(s, b, n)
-- 
2.51.0
